### DISCUSSION TITLE
How to perform k-steps of gradient updates.

### DISCUSSION THREAD

**Discussion by MichaelMMeskhi** [jax-ml/jax#13364] [Q&A] [answered]
In the flax documentation a training step is done as:

```python
@jax.jit
def train_step(state, batch):
  """Train for a single step."""
  def loss_fn(params):
    logits = CNN().apply({'params': params}, batch['image'])
    loss = cross_entropy_loss(logits=logits, labels=batch['label'])
    return loss, logits
  grad_fn = jax.grad(loss_fn, has_aux=True)
  grads, logits = grad_fn(state.params)
  state = state.apply_gradients(grads=grads)
  metrics = compute_metrics(logits=logits, labels=batch['label'])
  return state, metrics
```

This constitutes for 1 gradient step. Would it be correct to add a simple for loop to perform k-steps as:

```python
@jax.jit
def train_step(state, batch):
  """Train for a single step."""
  def loss_fn(params):
    logits = CNN().apply({'params': params}, batch['image'])
    loss = cross_entropy_loss(logits=logits, labels=batch['label'])
    return loss, logits

  for i in range(k):
      grad_fn = jax.grad(loss_fn, has_aux=True)
      grads, logits = grad_fn(state.params)
      state = state.apply_gradients(grads=grads)

  metrics = compute_metrics(logits=logits, labels=batch['label'])
  return state, metrics
```

In my case, I do not do batch training and instead utilize the entire dataset ~3600 data points.

**Comment by pharringtonp19** [comment, marked as answer]
If you are interested in compiling the entire training loop, my understanding is that you are better off using `jax.lax.scan`. 

Below is an example of how this might look (I didn't run this code, so an error may exist)
```
epochs = 1000 
init_state = ...
batch = ...

def loss_fn(params, batch):
    logits = CNN().apply({'params': params}, batch['image'])
    loss = cross_entropy_loss(logits=logits, labels=batch['label'])
    return loss, logits

def update(state):
    grads, logits = jax.grad(loss_fn, has_aux=True)(state.params, batch)
    state = state.apply_gradients(grads=grads)
    metrics = compute_metrics(logits=logits, labels=batch['label'])
    return state, metrics

@partial(jax.jit, static_argnums=(1,))
def train(init_state, epochs):
    opt_state, metrics = jax.lax.scan(update, init_state, xs=none, length=epochs)
    return opt_state, metrics

**Reply by MichaelMMeskhi** [reply]
Oh this is interesting approach. Thank you!
